File: dotnet-mvc/CleanArchitecture/UseCases/ManageStockUseCase.cs
using CleanArchitecture.Entities;
using CleanArchitecture.UseCases.Ports;

namespace CleanArchitecture.UseCases;

/// <summary>
/// Use Case: Manage product stock (increase/decrease).
/// This belongs to the USE CASES layer (application business rules).
/// The entity enforces the enterprise rule that stock cannot go negative.
/// </summary>
public class ManageStockUseCase
{
    private readonly IProductGateway _gateway;

    public ManageStockUseCase(IProductGateway gateway)
    {
        _gateway = gateway;
    }

    public async Task<Product> DecreaseStock(Guid id, int quantity)
    {
        var product = await _gateway.FindById(id);
        if (product is null)
            throw new ProductNotFoundException(id);

        // Entity enforces enterprise rule: stock cannot go below zero
        product.DecreaseStock(quantity);

        return await _gateway.Update(product);
    }

    public async Task<Product> IncreaseStock(Guid id, int quantity)
    {
        var product = await _gateway.FindById(id);
        if (product is null)
            throw new ProductNotFoundException(id);

        product.IncreaseStock(quantity);

        return await _gateway.Update(product);
    }
}

File: java-mvc/clean-architecture/src/main/java/com/ecommerce/clean/usecase/ManageStockUseCase.java
package com.ecommerce.clean.usecase;

import com.ecommerce.clean.entity.Product;
import com.ecommerce.clean.usecase.port.ProductGateway;

import java.util.UUID;

/**
 * USE CASE: Manage product stock (increase/decrease).
 */
public class ManageStockUseCase {

    private final ProductGateway productGateway;

    public ManageStockUseCase(ProductGateway productGateway) {
        this.productGateway = productGateway;
    }

    public void decrease(UUID productId, int quantity) {
        Product product = productGateway.findById(productId)
                .orElseThrow(() -> new IllegalArgumentException("Product not found: " + productId));
        product.decreaseStock(quantity);
        productGateway.save(product);
    }

    public void increase(UUID productId, int quantity) {
        Product product = productGateway.findById(productId)
                .orElseThrow(() -> new IllegalArgumentException("Product not found: " + productId));
        product.increaseStock(quantity);
        productGateway.save(product);
    }
}
